packages/x-scheduler/src/primitives/use-event-calendar/store.ts

```diff
@@ -213,6 +212,64 @@ export const selectors = {
       });
     },
   ),
+  eventsToRenderGroupedByResource: createSelector(
+    (state: State) => state.events,
+    (state: State) => state.adapter,
+    (events, adapter, { start, end }) => {
+      const groupedEvents: Record<string, CalendarEvent[][]> = {};
+      const occupiedRows: Record<string, { start: DateTime; end: DateTime }[][]> = {};
+
+      for (const event of events) {
+        let rowIndex = 0;
+        const resourceId = event.resource;
+
+        if (adapter.isAfter(event.start, end) || adapter.isBefore(event.end, start)) {
+          continue;
+        }
+
+        if (resourceId) {
+          if (!groupedEvents[resourceId]) {
+            groupedEvents[resourceId] = [[]];
+          }
+
+          let i = 0;
+          while (i < groupedEvents[resourceId].length) {
+            if (groupedEvents[resourceId][i].length === 0) {
+              break;
+            }
+
+            let hasConflict = groupedEvents[resourceId][i].find((placedEvent) => {
+              const overlaps =
+                adapter.isBefore(event.start, placedEvent.end) &&
+                adapter.isAfter(event.end, placedEvent.start);
+
+              const sameInterval =
+                adapter.toJsDate(event.start).getTime() ===
+                  adapter.toJsDate(placedEvent.start).getTime() &&
+                adapter.toJsDate(event.end).getTime() ===
+                  adapter.toJsDate(placedEvent.end).getTime();
+
+              return overlaps || sameInterval;
+            });
+
+            if (hasConflict) {
+              i += 1;
+            } else {
+              break;
+            }
+          }
+
+          rowIndex = i;
+          if (rowIndex >= groupedEvents[resourceId].length) {
+            groupedEvents[resourceId].push([]);
```

Comment:
this would result in something like this

<img width="1779" height="635" alt="Screenshot 2025-09-04 150952" src="https://github.com/user-attachments/assets/9ab7ebc5-9146-4714-8310-df179fdcc39d" />